### ISSUE TITLE
ENOMEM issue when using design system rules in wsl

### ISSUE THREAD

**Issue by TheRealSyler** [sergioazoc/oxlint-tailwindcss#24]
I get the following error when debug mode is enabled

```sh
[oxlint-tailwindcss] Failed to load design system from "/home/syler/GitHub/morion/src/styles.css": spawnSync /home/syler/.nvm/versions/node/v25.9.0/bin/node ENOMEM
```

Im using vite plus in wsl, there is not much more information i can give. I asked gpt-5.5 to find and fix the issue, the fix works but i didn't open a pr because i have not fully understood the fix. You can take a look at gpts solution [here](https://github.com/sergioazoc/oxlint-tailwindcss/compare/main...TheRealSyler:oxlint-tailwindcss:main).

According to gpt the problem is:

> **Root cause:** On a cold disk cache, oxlint’s parallel isolates each called `loadDesignSystemSync`, which used `execFileSync` to fork a new Node process. Several of those forks at once caused `spawnSync ... ENOMEM` on WSL/Linux.

My config:

```js
      "tailwindcss/no-unknown-classes": "error",
      "tailwindcss/no-duplicate-classes": "error",
      "tailwindcss/no-conflicting-classes": "error",
      "tailwindcss/no-deprecated-classes": "error",
      "tailwindcss/no-unnecessary-whitespace": "error",
      "tailwindcss/no-dark-without-light": "warn",
      "tailwindcss/no-contradicting-variants": "warn",
      "tailwindcss/enforce-canonical": "warn",
      "tailwindcss/enforce-sort-order": "warn",
      "tailwindcss/enforce-shorthand": "warn",
      "tailwindcss/enforce-logical": "error",
      "tailwindcss/enforce-consistent-important-position": "warn",
      "tailwindcss/enforce-negative-arbitrary-values": "warn",
      "tailwindcss/enforce-consistent-variable-syntax": "warn",
      "tailwindcss/consistent-variant-order": "warn",
      "tailwindcss/max-class-count": "error",
      "tailwindcss/enforce-consistent-line-wrapping": [
        "error",
        {
          printWidth: 200,
        },
      ],
      "tailwindcss/no-restricted-classes": "error",
      "tailwindcss/no-arbitrary-value": "off",
      "tailwindcss/no-hardcoded-colors": "warn",
      "tailwindcss/no-unnecessary-arbitrary-value": "warn",
      "tailwindcss/prefer-theme-tokens": "error",
```

Let me know if there is more info i can give you, i can make a repro repo tomorrow if you need it.

**Comment by sergioazoc** [comment]
Fixed — thanks @TheRealSyler, and your root-cause analysis was spot on.

On a cold disk cache, oxlint's parallel isolates each forked their own `execFileSync` precompute child for the same CSS, and a dozen simultaneous forks (each loading `@tailwindcss/node`) exhausted memory — exactly the `spawnSync … ENOMEM` you hit.

v1.0.0 wraps the fork in a content-hash-scoped file lock: the first isolate computes and writes the cache, the rest wait (an Atomics-based sync sleep) for the resulting JSON instead of forking. A stale lock — older than the precompute timeout plus a write margin, i.e. a holder that died mid-compute — is reclaimed so a crash can't wedge every isolate, and a non-writable temp dir degrades to an uncoordinated compute rather than spinning. Coordination is keyed by content hash, so distinct CSS files still precompute in parallel.

Shipping in v1.0.0 (releasing shortly). Closing as fixed — please reopen if it still bites on WSL.

**Comment by TheRealSyler** [comment]
I just tried the new version, the error is a lot nicer (and i don't need to enable debug mode to see it), but i still get the same error:

`
  × tailwindcss(prefer-theme-tokens): Failed to precompute design system from "/home/syler/GitHub/morion/src/styles.css": spawnSync /home/syler/.nvm/versions/node/v25.9.0/bin/node ENOMEM
`

(i don't have the right permissions to reopen the issue)

**Comment by sergioazoc** [comment]
Thanks for testing the new version and reporting back.

What you're hitting now is a different failure mode than the one this issue tracked. The original report was a **fork storm**: every parallel oxlint isolate forked its own precompute child for the same CSS at once, and a dozen simultaneous forks exhausted memory. v1.0.0's content-hash lock fixed that — only one isolate forks now, the rest wait for the cache file. That part is working (hence the cleaner, single diagnostic instead of a flood).

The residual `ENOMEM` is the cost of that **single** fork, not concurrency. `execFileSync` does `fork()` + immediate `exec()`. On Linux the child never writes the parent's copy-on-write pages — it replaces them with `node` instantly — so the memory the fork "needs" is never actually touched. It's a purely accounting-level reservation the kernel demands under its default overcommit heuristic. When the parent (the oxlint host with its isolates resident) is large and the box is memory-constrained, that bookkeeping check refuses the fork even though no real RAM would be used. This is the same situation Redis warns about with its `vm.overcommit_memory` notice.

Because it's environmental rather than a concurrency bug, the fix is on the host side:

- **`sudo sysctl vm.overcommit_memory=1`** (persist via `/etc/sysctl.conf`). This is almost certainly the one-liner that unblocks you — it tells the kernel not to do the accounting reservation for a fork-then-exec.
- **Add swap / raise the memory cap in `.wslconfig`** (`[wsl2]` → `memory=…`, `swap=…`). WSL2 defaults are conservative and often ship without usable swap.
- **Lower oxlint's parallelism** (e.g. `--threads`) so the parent process is smaller at fork time.

I'd start with `vm.overcommit_memory=1` and report back. WSL on a memory-constrained host is an uncommon setup and this is the only report we've had, so I don't want to bolt an external broker process onto the plugin to work around a kernel overcommit setting — that would cut against the "only `@tailwindcss/node` + `tailwindcss`, no external workers" design. If the overcommit toggle doesn't resolve it for you, let us know here and we'll reconsider.

**Comment by TheRealSyler** [comment]
I tested the latest version now it works (without needing to configure my WSL), thank you for fixing it.
